Add `dump-co-pilot-reporting-data.sh`

iaindillingham opened the pull request:
After completing #3958, Millie asked for a dump of co-pilot reporting data. Whilst the snippet in `DEVELOPERS.md` is almost complete, I realized I needed to remove a pre-existing dump before running the snippet and zip the dump before sharing it. For convenience, then, I decided to move the snippet into a script.

ghickman commented on scripts/dump-co-pilot-reporting-data.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Can you use `#!/usr/bin/env bash` here?  It's a little more portable and how we do it elsewhere.

iaindillingham replied:
I can, but each of the other scripts in `scripts` uses `#!/bin/bash` rather than `#!/usr/bin/env bash`. I agree that the latter is more portable. Should I change them all?

ghickman replied:
Oh yes, that would be great.  IIRC @madwort normalised all our justfile bash invocations to the env-based one, so it would be good to be consistent.

iaindillingham replied:
I've changed them all.

ghickman commented on scripts/dump-co-pilot-reporting-data.sh:
```diff
@@ -0,0 +1,32 @@
+    exit 1
+fi
+
+if [ -z ${DATABASE_URL+none} ]; then
```
What does `+none` do here?

iaindillingham replied:
If `DATABASE_URL` isn't set, then the braces will evaluate to an empty string. If it is set, then the braces will evaluate to the string `none`. In the former case, [the condition will be true][1] (length is zero). In the latter case, the condition will be false (length is four). I couldn't think of a better token. Possibly `false`? (More on [SO][2].)

Why bother? Well, without the braces, if `DATABASE_URL` isn't set, then the script terminates with

```
./scripts/dump-co-pilot-reporting-data.sh: line 9: DATABASE_URL: unbound variable
```

is displayed, because of `set -eu`.

[1]: https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html
[2]: https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash

ghickman replied:
Well that's no fun, is it?  @bloodearnest – is there a better way you know of to check for an env var existing and also having a value?

bloodearnest replied:
I could totally do a yarn on test vs [ vs [[, thinking about it.

iaindillingham replied:
I'd love a yarn on that! How about Thursday 25th January?

iaindillingham replied:
Thank you for your comments, @bloodearnest. Really useful and really interesting! I've updated the script.

ghickman reviewed (commented):
This is a great start, thank you for improving the state of this task!

Could you move the script into a just target?  We've worked hard to keep entry points inside just on the principle of least surprise.  It also means you can add a little docstring to the target which shows up when you run `just`.

If you think it's a little long for that you could always invoke the script from just like we do for several other things.

iaindillingham commented:
> Could you move the script into a just target? We've worked hard to keep entry points inside just on the principle of least surprise. It also means you can add a little docstring to the target which shows up when you run `just`.
> 
> If you think it's a little long for that you could always invoke the script from just like we do for several other things.

Thanks for the high quality feedback, @ghickman and @bloodearnest. George, I've added a Just target that points to the script. I've also made a couple of changes to the script since you last looked at it. As well as Simon's suggestions, it now removes the uncompressed SQLite database. I've also added the compressed SQLite database to `.gitignore`, so we don't accidentally commit it.

ghickman reviewed (approved):
This looks great, thanks Iain 👍

bloodearnest commented on scripts/dump-co-pilot-reporting-data.sh:
```diff
@@ -0,0 +1,33 @@
+    --table jobserver_releasefile \
+    ${DATABASE_URL} \
+    jobserver.sqlite
+
```
Minor suggestion - it might be worth running sqlite's VACCUUM command to shrink the db file before then compressing.

```sqlite3 jobserver.sqlite VACUUM```

This might not make much difference, depending on how db-to-sqlite does its thing, however.

iaindillingham replied:
Thanks, Simon. I tried this, and it didn't make any difference.
